### PR TITLE
Add torch.jit.Const

### PR DESCRIPTION
**Pull request by zdevito**
`Const` allows you to mark certain python attributes of a modules constant,
enabling script methods to use them in way that are otherwise not allowed.

* This includes using python constants like `use_bias` in if statements.

* Add support for unrolledFor to torch.jit.Const

#6024

### CODE REVIEW

**Comment by zdevito** [comment]
@apaszke

**Comment by ezyang** [comment]
This contains a spurious update to the cpuinfo submodule

File: torch/csrc/jit/script/compiler.cpp [hunk 1]
```diff
@@ -477,77 +477,74 @@ struct to_ir {
...
-    auto target_list = List<Ident>(targets);
-    if (target_list.size() != 1) {
-      throw ErrorReport(stmt)
-          << "Iteration variable unpacking is not supported";
+    // it isn't a range(<expr>) loop, treat it as a sugared value that maybe can be
+    // unrolled
+    auto sv = emitSugaredExpr(itrs[0]);
+    auto instances = sv->unrolledFor(stmt.range(), method);
+    for(auto inst : instances) {
+      environment_stack->setSugaredVar(targets[0].name(), inst);
```
Comment (apaszke) [inline torch/csrc/jit/script/compiler.cpp]: This will leave the iterator var in the environment, which is inconsistent with other kinds of loops IIRC.

File: torch/csrc/jit/script/compiler.h [hunk 2]
```diff
@@ -44,6 +44,14 @@ struct SugaredValue : public std::enable_shared_from_this<SugaredValue> {
     size_t n_outputs) {
     throw ErrorReport(loc) << "cannot call a " << kind();
   }
+
+  // use it in `for i in this: ...`
+  // in this case we will unroll the loop boy by assigning i to each of
```
Comment (apaszke) [inline torch/csrc/jit/script/compiler.h]: typo: body

File: torch/csrc/jit/script/init.cpp [hunk 3]
```diff
@@ -83,6 +88,46 @@ struct VISIBILITY_HIDDEN PythonValue : public SugaredValue {
...
+// 1. conversion via asValue to a constant Tensor
+// 2. unrolling of for loops
+struct VISIBILITY_HIDDEN ConstantPythonValue : public PythonValue {
+  using PythonValue::PythonValue;
+  virtual Value * asValue(SourceRange loc, Method & m) override {
+    if(py::isinstance<py::int_>(self)) {
+      return createConstant(loc, m, at::CPU(at::kInt).scalarTensor(py::cast<int32_t>(self)));
+    } else if(py::isinstance<py::float_>(self)) {
+      return createConstant(loc, m, at::CPU(at::kFloat).scalarTensor(py::cast<float>(self)));
+    } else if(py::isinstance<py::bool_>(self)) {
```
Comment (apaszke) [inline torch/csrc/jit/script/init.cpp]: why the underscores?

File: torch/csrc/jit/script/init.cpp [hunk 4]
```diff
@@ -83,6 +88,46 @@ struct VISIBILITY_HIDDEN PythonValue : public SugaredValue {
...
+      return createConstant(loc, m, at::CPU(at::kByte).scalarTensor(py::cast<bool>(self)));
```
Comment (apaszke) [inline torch/csrc/jit/script/init.cpp]: We really should have a single function like `createBool`, `createFloat` to unify how we create primitive values in the IR. It will make maintenance much easier in case we'll ever want to change what we do

File: torch/csrc/jit/script/init.cpp [hunk 5]
```diff
@@ -161,6 +210,14 @@ struct ModuleValue : public SugaredValue {
   std::shared_ptr<Module> module;
 };
 
+static std::shared_ptr<SugaredValue> createConstantSugaredValue(py::object obj) {
+  if(py::isinstance<Module>(obj)) {
+    auto r = py::cast<std::shared_ptr<Module>>(obj);
+    return std::make_shared<ModuleValue>(r);
+  }
```
Comment (apaszke) [inline torch/csrc/jit/script/init.cpp]: Why do we need the special case for `Module`? Also, it won't make the module immutable, like it does with other Python objects.

NB `Module` here is the C++ module (i.e. only `ScriptedModule` in Python), right?

File: torch/jit/__init__.py [hunk 6]
```diff
@@ -656,6 +656,25 @@ def __getitem__(self, k):
...
+        return v
+    elif isinstance(v, tuple) or isinstance(v, list):
+        return tuple(_get_valid_constant(x) for x in v)
+    constants = ", ".join(typ.__name__ for typ in _constant_types)
+    raise TypeError(
+        "'{}' object is not a valid constant or a list/tuple of valid constants. ".format(type(v).__name__) +
+        "Valid constants are {{{}}}".format(constants))
+
+
+class Const(object):
```
Comment (apaszke) [inline torch/jit/__init__.py]: I know this might be bikeshedding, but can we please make a function for creating constants and call it `const` (and rename this to `_Const` or sth like that). I think it would be more consistent with other API we expose, and wouldn't mean "create this class", but only "mark this as const" (somehow, doesn't matter how).

File: torch/jit/__init__.py [hunk 7]
```diff
@@ -656,6 +656,25 @@ def __getitem__(self, k):
             raise KeyError(k)
         return self.module._get_parameter(k)
 
+# types that can be constants. If you edit this list,
+# then you also need to edit the handlers in ConstantValue in jit/script/init.cpp
+_constant_types = [bool, float, int, Module, types.FunctionType]
+
+
+def _get_valid_constant(v):
+    if any(isinstance(v, typ) for typ in _constant_types):
```
Comment (apaszke) [inline torch/jit/__init__.py]: you can pass a tuple of instances to `isinstance`: `isinstance(v, _constant_types)`

File: torch/jit/__init__.py [hunk 8]
```diff
@@ -656,6 +656,25 @@ def __getitem__(self, k):
...
+        return v
+    elif isinstance(v, tuple) or isinstance(v, list):
+        return tuple(_get_valid_constant(x) for x in v)
```
Comment (apaszke) [inline torch/jit/__init__.py]: It would be useful to note that `_constant_types` isn't an exhaustive list, and should not be used to implement checks

File: torch/jit/__init__.py [hunk 9]
```diff
@@ -656,6 +656,25 @@ def __getitem__(self, k):
             raise KeyError(k)
         return self.module._get_parameter(k)
 
+# types that can be constants. If you edit this list,
+# then you also need to edit the handlers in ConstantValue in jit/script/init.cpp
+_constant_types = [bool, float, int, Module, types.FunctionType]
```
Comment (apaszke) [inline torch/jit/__init__.py]: Do you need to mark `nn.Module` as constant to be able to use it inside scripted functions? Also, it doesn't really mean that the module (as in its state, attributes, etc.) will remain constant, right? Only that this attribute won't be re-assigned later

File: torch/jit/__init__.py [hunk 10]
```diff
@@ -720,6 +758,29 @@ def define(self, lang):
...
+        # fast check, we already know about this module
+        # and do not need to recalculate it
+        if id(m) in self._all_script_modules:
+            return True
+        # slow check, recalculate all submodules (more may have been added)
+        return id(m) in self._calc_all_script_modules()
+
+    def _calc_all_script_modules(self):
+        for m in self._modules.values():
+            if(isinstance(m, ScriptModule)):
```
Comment (apaszke) [inline torch/jit/__init__.py]: please don't use parenthesis with `if` statements

File: torch/jit/__init__.py [hunk 11]
```diff
@@ -720,6 +758,29 @@ def define(self, lang):
         rcb = createResolutionCallback()
         self._define(lang, rcb, True)
 
+    def _check_modules_are_owned(self, v):
+        if isinstance(v, ScriptModule) and not self._is_script_submodule(v):
+            raise RuntimeError("torch.jit.Const includes a module '{}'".format(type(v).__name__) +
+                               " which is not a submodule of this module")
```
Comment (apaszke) [inline torch/jit/__init__.py]: Can't we just disallow assigning scripted modules as const variables? This seems to be a lot simpler, and I don't think we get a lot of value from what we have here (since it had to be assigned before anyway)

File: test/test_jit.py [hunk 12]
```diff
@@ -2021,6 +2021,132 @@ def forward(self, input):
...
+                return self.weight + thing
+
+        class M(torch.jit.ScriptModule):
+            def __init__(self):
+                super(M, self).__init__(False)
+                mods = []
+                for i in range(10):
+                    m = Sub()
+                    self.add_module('m' + str(i), m)
+                    mods.append(m)
```
Comment (apaszke) [inline test/test_jit.py]: Keeping modules in regular Python lists is a very bad idea. It will break e.g. with `DataParallel` and in multiple other occasions. That's why we have `nn.ModuleList`

**Review by apaszke — COMMENTED** [review]
#### Warning: GitHub hides 2 comments. Please remember to take a look at them too.

The implementation is fine (although I left some comments), but I really have mixed feelings about this interface. Take this for an example:
```python
self.attr = torch.jit.const(value)
```
If I were to look at this line without really understanding what `torch.jit.const` is, I'd just assume that it somehow "constifies" the value, and stores it in a regular attr. However, the scope of `const` really encompasses the whole line, because from that point on this attribute will really be treated specially (even more than its value!), so it's really (not valid Python ofc)
```python
torch.jit.value(self.attr = value)
# or
torch.jit.will_be_const_after_init(self.attr)
self.attr = value
```
or, using Python3 type annotation syntax
```python
self.attr: torch.jit.const = value
```

I don't have a proposal in mind right now, but it would be nice to make the meaning of such assignment much more explicit (because e.g. `var = const(...)` doesn't have the same power!!)

File: torch/jit/__init__.py (hunk 9 above)
Comment (zdevito) [reply torch/jit/__init__.py]: No you don't need to mark a module const, but you do want to label a _list_ of modules const. Maybe this should be more integrated with ModuleList, I will look into how it works.

File: torch/csrc/jit/script/init.cpp (hunk 5 above)
Comment (zdevito) [reply torch/csrc/jit/script/init.cpp]: Calls from ScriptModule->ScriptModule need to be handled differently than calls from ScriptModule->python module. So that is what is happening here.

File: torch/csrc/jit/script/init.cpp (hunk 3 above)
Comment (zdevito) [reply torch/csrc/jit/script/init.cpp]: I guess because float and bool are keywords

File: test/test_jit.py (hunk 12 above)
Comment (zdevito) [reply test/test_jit.py]: In this case, they are both added as a submodule via add_module, and added to the list. Is that still a problem?

Comment (apaszke) [reply test/test_jit.py]: Yes. The list can get out of sync with `_modules`.

File: torch/csrc/jit/script/init.cpp (hunk 5 above)
Comment (apaszke) [reply torch/csrc/jit/script/init.cpp]: Sorry I didn't get that. Can you please expand your comment?

**Comment by zdevito** [comment]
I don't have a better idea for `const` since there are a few constraints:

* We want to limit const things to attributes, otherwise we really can't stop people from re-assigning the value of a const 
```
a = const(3)
....
a = const(4) # oops, we can't prevent this, but people might thing it should work
```

* We need to ensure ScriptModule `parent` that calls a ScriptModule `child` actually owns child, otherwise we violate the invariant that ScriptModules can only call modules that are their children.

But I am open to other suggestions.

File: test/test_jit.py (hunk 12 above)
Comment (zdevito) [reply test/test_jit.py]: We can have torch.jit.const([<module_list>]) return a 'ConstModuleList' which would be a subclass of ScriptModule that also behaves like a list similar to nn.ModuleList. This would then eliminate all the "is this a submodule" checking, which I like.

Comment (apaszke) [reply test/test_jit.py]: Sounds good!

**Comment by ezyang** [comment]
One silly alternate interface which works around @apaszke's concerns is something like setconstattr(mod, 'foo', v). Easy to polyfill Python side and emphasizes the fact that you are const-ifying the attribute.

File: torch/csrc/jit/script/init.cpp (hunk 5 above)
Comment (zdevito) [reply torch/csrc/jit/script/init.cpp]: This special case will be moved into the handling for nn.ModuleList. We need to make script modules into ModuleValues and normal modules into PythonValues because we need to ensure calls to script modules do not go through PythonValues. Otherwise, instead of inlining the graph, it would insert a python op into the graph to call the subgraph.

**Comment by zdevito** [comment]
Rather than use `torch.jit.const` this new update uses a statically declared list of constant attributes to distinguish when the script module should treat a property as a constant and enforce its const-ness:

```
class Foo(ScriptModule):
     __constants__ = ['a', 'b', 'c']
```

**Comment by zdevito** [comment]
@apaszke This should be ready to go, but I need an approve to merge.

File: test/test_jit.py [hunk 13]
```diff
@@ -2061,6 +2061,147 @@ def forward(self, inp):
...
+
+            @torch.jit.script_method
+            def forward(self):
+                return self.b, self.i, self.c
+
+        m = M()
+        o0, o1, o2 = m()
+        self.assertEqual(o0, 0)
+        self.assertEqual(o1, 1)
+        self.assertEqual(o2, 3.5)
```
Comment (apaszke) [inline test/test_jit.py]: Will those be actual Python numbers or PyTorch scalars?

File: torch/csrc/jit/script/init.cpp [hunk 14]
```diff
@@ -157,10 +202,29 @@ struct ModuleValue : public SugaredValue {
...
+    if(!py::isinstance(py_module, py::module::import("torch.jit").attr("_ConstModuleList")))
+      return SugaredValue::unrolledFor(loc, m);
+    std::vector<std::shared_ptr<SugaredValue>> result;
+    for(py::handle module : py_module) {
+      py::object obj = py::reinterpret_borrow<py::object>(module);
+      if(py::isinstance<Module>(obj)) {
+        auto r = py::cast<std::shared_ptr<Module>>(obj);
+        result.push_back(std::make_shared<ModuleValue>(r));
+      } else {
+        result.push_back(std::make_shared<ConstantPythonValue>(obj));
```
Comment (apaszke) [inline torch/csrc/jit/script/init.cpp]: How would it be possible to get a constant here? Isn't this supposed to be a Module-only iterator?

**Review by apaszke — APPROVED** [review]
Looks great now! I left two minor comments.

File: torch/csrc/jit/script/init.cpp (hunk 14 above)
Comment (zdevito) [reply torch/csrc/jit/script/init.cpp]: It can be a non-script module, which we use a PythonValue for.

File: test/test_jit.py (hunk 13 above)
Comment (zdevito) [reply test/test_jit.py]: They are PyTorch scalars, I think the assertEqual is just smart enough to do the comparison.
